### ISSUE TITLE
Failed to run the benchmark_model

### ISSUE THREAD

**Issue by tankienleong** [tensorflow/tensorflow#13011]
System information
- Have I written custom code: No 
- OS Platform and Distribution: Linux Ubuntu 14.04
- TensorFlow installed from: source
- TensorFlow version: ('v1.3.0-rc1-1951-g04c318b', '1.3.0')
- Python version: 2.7.12
- Bazel version (if compiling from source): 0.5.0
- CUDA/cuDNN version: 8.0/5.1
- GPU model and memory: Quadro M4000 8Gb
Exact command to reproduce: bazel-bin/tensorflow/tools/benchmark/benchmark_model   --graph=inception5h/tensorflow_inception_graph.pb   --input_layer="input:0"   --input_layer_shape="1,224,224,3"   --input_layer_type="float" --output_layer="output:0" --show_run_order=false --show_time=false --show_memory=false --show_summary=false --show_flops=true

Problem
I follow the instruction on https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/benchmark to run the benchmark_model on desktop but it failed when I set the --show_flops=true. The tensorflow_inception_graph.pb is downloaded from the link provided on the page. 

Logs
(my_project) ubuntu@ubuntu-desktop:~/Desktop/virtualroiflow/tensorflow$ bazel-bin/tensorflow/tools/benchmark/benchmark_model   --graph=inception5h/tensorflow_inception_graph.pb   --input_layer="input:0"   --input_layer_shape="1,224,224,3"   --input_layer_type="float" --output_layer="output:0" --show_run_order=false --show_time=false --show_memory=false --show_summary=false --show_flops=true
2017-09-13 17:53:21.911797: I tensorflow/tools/benchmark/benchmark_model.cc:426] Graph: [inception5h/tensorflow_inception_graph.pb]
2017-09-13 17:53:21.911837: I tensorflow/tools/benchmark/benchmark_model.cc:427] Input layers: [input:0]
2017-09-13 17:53:21.911841: I tensorflow/tools/benchmark/benchmark_model.cc:428] Input shapes: [1,224,224,3]
2017-09-13 17:53:21.911844: I tensorflow/tools/benchmark/benchmark_model.cc:429] Input types: [float]
2017-09-13 17:53:21.911846: I tensorflow/tools/benchmark/benchmark_model.cc:430] Output layers: [output:0]
2017-09-13 17:53:21.911852: I tensorflow/tools/benchmark/benchmark_model.cc:431] Num runs: [1000]
2017-09-13 17:53:21.911870: I tensorflow/tools/benchmark/benchmark_model.cc:432] Inter-inference delay (seconds): [-1.0]
2017-09-13 17:53:21.911873: I tensorflow/tools/benchmark/benchmark_model.cc:433] Inter-benchmark delay (seconds): [-1.0]
2017-09-13 17:53:21.911877: I tensorflow/tools/benchmark/benchmark_model.cc:435] Num threads: [-1]
2017-09-13 17:53:21.911894: I tensorflow/tools/benchmark/benchmark_model.cc:436] Benchmark name: []
2017-09-13 17:53:21.911897: I tensorflow/tools/benchmark/benchmark_model.cc:437] Output prefix: []
2017-09-13 17:53:21.911900: I tensorflow/tools/benchmark/benchmark_model.cc:438] Show sizes: [0]
2017-09-13 17:53:21.911917: I tensorflow/tools/benchmark/benchmark_model.cc:439] Warmup runs: [2]
2017-09-13 17:53:21.911920: I tensorflow/tools/benchmark/benchmark_model.cc:54] Loading TensorFlow.
2017-09-13 17:53:21.911941: I tensorflow/tools/benchmark/benchmark_model.cc:61] Got config, 0 devices
2017-09-13 17:53:21.912358: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2017-09-13 17:53:22.022462: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:892] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-09-13 17:53:22.022832: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Found device 0 with properties: 
name: Quadro M4000 major: 5 minor: 2 memoryClockRate(GHz): 0.7725
pciBusID: 0000:01:00.0
totalMemory: 7.93GiB freeMemory: 7.61GiB
2017-09-13 17:53:22.022864: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1055] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: Quadro M4000, pci bus id: 0000:01:00.0, compute capability: 5.2)
2017-09-13 17:53:22.078421: E tensorflow/core/framework/op_kernel.cc:1142] OpKernel ('op: "Fact" device_type: "CPU" label: "sergey"') for unknown op: Fact
2017-09-13 17:53:22.078460: E tensorflow/core/framework/op_kernel.cc:1142] OpKernel ('op: "Fact" device_type: "CPU" label: "Sergey"') for unknown op: Fact
2017-09-13 17:53:22.078466: E tensorflow/core/framework/op_kernel.cc:1142] OpKernel ('op: "Fact" device_type: "GPU" host_memory_arg: "fact"') for unknown op: Fact
2017-09-13 17:53:22.081269: I tensorflow/tools/benchmark/benchmark_model.cc:291] Running benchmark for max 2 iterations, max -1 seconds without detailed stat logging, with -1s sleep between inferences
2017-09-13 17:53:22.173360: I tensorflow/tools/benchmark/benchmark_model.cc:324] count=2 first=59501 curr=32424 min=32424 max=59501 avg=45962.5 std=13538

2017-09-13 17:53:22.173384: I tensorflow/tools/benchmark/benchmark_model.cc:291] Running benchmark for max 1000 iterations, max 10 seconds without detailed stat logging, with -1s sleep between inferences
2017-09-13 17:53:32.199487: I tensorflow/tools/benchmark/benchmark_model.cc:324] count=316 first=33259 curr=33929 min=29881 max=34487 avg=31683.6 std=684

2017-09-13 17:53:32.199512: I tensorflow/tools/benchmark/benchmark_model.cc:291] Running benchmark for max 1000 iterations, max 10 seconds with detailed stat logging, with -1s sleep between inferences
2017-09-13 17:53:32.200083: I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libcupti.so.8.0 locally
2017-09-13 17:53:42.310948: I tensorflow/tools/benchmark/benchmark_model.cc:324] count=316 first=51892 curr=32599 min=29882 max=51892 avg=31710.3 std=1338

2017-09-13 17:53:42.310986: I tensorflow/tools/benchmark/benchmark_model.cc:538] Average inference timings in us: Warmup: 45962, no stats: 31683, with stats: 31710
2017-09-13 17:53:42.311006: I tensorflow/core/util/stat_summarizer.cc:358] Number of nodes executed: 141
2017-09-13 17:53:42.311157: I tensorflow/core/util/stat_summarizer.cc:468] ============================== Summary by node type ==============================
2017-09-13 17:53:42.311162: I tensorflow/core/util/stat_summarizer.cc:468] 	             [Node type]	  [count]	  [avg ms]	    [avg %]	    [cdf %]	  [mem KB]	[times called]
2017-09-13 17:53:42.311166: I tensorflow/core/util/stat_summarizer.cc:468] 	                  Conv2D	       22	    35.377	    76.119%	    76.119%	 10077.888	       22
2017-09-13 17:53:42.311170: I tensorflow/core/util/stat_summarizer.cc:468] 	                     LRN	        2	     4.296	     9.243%	    85.362%	  3211.264	        2
2017-09-13 17:53:42.311173: I tensorflow/core/util/stat_summarizer.cc:468] 	                 MaxPool	        6	     3.124	     6.722%	    92.084%	  3562.496	        6
2017-09-13 17:53:42.311177: I tensorflow/core/util/stat_summarizer.cc:468] 	                 BiasAdd	       24	     1.677	     3.608%	    95.692%	     0.000	       24
2017-09-13 17:53:42.311195: I tensorflow/core/util/stat_summarizer.cc:468] 	                    Relu	       23	     0.740	     1.592%	    97.285%	     0.000	       23
2017-09-13 17:53:42.311199: I tensorflow/core/util/stat_summarizer.cc:468] 	                  MatMul	        2	     0.703	     1.513%	    98.797%	     8.128	        2
2017-09-13 17:53:42.311216: I tensorflow/core/util/stat_summarizer.cc:468] 	                  Concat	        3	     0.375	     0.807%	    99.604%	  2706.368	        3
2017-09-13 17:53:42.311220: I tensorflow/core/util/stat_summarizer.cc:468] 	                   Const	       51	     0.082	     0.176%	    99.781%	     0.000	       51
2017-09-13 17:53:42.311225: I tensorflow/core/util/stat_summarizer.cc:468] 	                 AvgPool	        1	     0.057	     0.123%	    99.903%	    32.512	        1
2017-09-13 17:53:42.311229: I tensorflow/core/util/stat_summarizer.cc:468] 	                 Softmax	        1	     0.028	     0.060%	    99.963%	     0.000	        1
2017-09-13 17:53:42.311232: I tensorflow/core/util/stat_summarizer.cc:468] 	                    NoOp	        1	     0.006	     0.013%	    99.976%	     0.000	        1
2017-09-13 17:53:42.311236: I tensorflow/core/util/stat_summarizer.cc:468] 	                    _Arg	        1	     0.004	     0.009%	    99.985%	     0.000	        1
2017-09-13 17:53:42.311252: I tensorflow/core/util/stat_summarizer.cc:468] 	                 _Retval	        1	     0.003	     0.006%	    99.991%	     0.000	        1
2017-09-13 17:53:42.311256: I tensorflow/core/util/stat_summarizer.cc:468] 	                 Reshape	        2	     0.003	     0.006%	    99.998%	     0.000	        2
2017-09-13 17:53:42.311274: I tensorflow/core/util/stat_summarizer.cc:468] 	                Identity	        1	     0.001	     0.002%	   100.000%	     0.000	        1
2017-09-13 17:53:42.311278: I tensorflow/core/util/stat_summarizer.cc:468] 
2017-09-13 17:53:42.416261: E tensorflow/tools/benchmark/benchmark_model.cc:556] FLOPs calculation failed with Invalid argument: You must feed a value for placeholder tensor 'input' with dtype float
	 [[Node: input = Placeholder[dtype=DT_FLOAT, shape=[], _device="/job:localhost/replica:0/task:0/cpu:0"]()]]

**Comment by reedwm** [comment]
@andrewharp can you fix this?

**Comment by tankienleong** [comment]
Hi @andrewharp , any update on this issue?

**Comment by tankienleong** [comment]
Problem solved after I modified the input argument.
`bazel-bin/tensorflow/tools/benchmark/benchmark_model --graph=inception5h/tensorflow_inception_graph.pb --input_layer="input" --input_layer_shape="1,224,224,3" --input_layer_type="float" --output_layer="output" --show_run_order=false --show_time=false --show_memory=false --show_summary=false --show_flops=true `

**Comment by fengyuICT** [comment]
well, the reason is:
1) in tensorflow/tools/benchmark/benchmark_model.cc, function GetOutputShapes will determining which graph node will be fecthed. code is list below:
`   for (const string& wanted_shape : wanted_shapes) {
      bool is_input = false;
      for (const std::pair<string, tensorflow::Tensor>& input_tensor :
           input_tensors) {                                                                
        if (input_tensor.first == wanted_shape) {
          (*node_shapes)[wanted_shape] = input_tensor.second.shape();
          is_input = true;
          break;
        }   
      }        
      input_count++;
      if (!is_input) {
        output_tensor_names.push_back(wanted_shape);
      }        
    }        `
wanted_shape include all ops that its type is ["Conv2D", "MatMul", "QuantizedConv2D", "QuantizedMatMul", "DepthwiseConv2dNative" ], if you pass "input:0" to input_layer then "input:0" not match wanted_shape or its input tensors(ops), so "input:0" will put into fetch_list. In session run, method RewriteGraphForExecution will be invoke to replace placeholder with real op.(see in tensorflow/core/graph/subgraph.cc). but that cause a problem since "input:0" is in fetch_list, the graph will change to that:
<Before rewrite>
_source --> "input:0" -> other op(e.g. Conv2D) --> _sink
<after rewrite>
_source --> "__arg_input"(type: _Arg) --> other op -> _sink
_source --> "input:0"(type:PlaceHolderOp) --> _sink
that cause a problem because no placeholder op exist in the runtime(Before Executor::Process).
However if you set input_layer="input",  the placeholderop is not exist in fetch_list, so the graph is changed to:
_source --> "input:0"(type:PlaceHolderOp)
that will optimized since this path not link to _sink.

**Comment by fengyuICT** [comment]
Another problem that confuses me is that although set input_layer="input:0" but it not cause a error at warmup phase, model execute phase(see in benckmark_model.cc). And I find the reason is fetch_nodes is exactly equal to output_layer.
So a patch can fixed this problem:
`
std::vector<string> placeholder_name;
Status CalculateFlops(...) {
  if (node.op().compare("Placeholder") == 0 ||
              node.op().compare("PlaceholderV2") == 0) {
    placeholder_name.push_back(node.name());
  }
  Status GetOutputShapes(const std::vector<InputLayerInfo>& inputs,
                         const std::set<string>& wanted_shapes, Session* session,
                         std::unordered_map<string, TensorShape>* node_shapes) {
    a   for (const string& wanted_shape : wanted_shapes) {
      bool is_input = false;                                                                                                                                                               
      for (const std::pair<string, tensorflow::Tensor>& input_tensor :
           input_tensors) {
        /..../
        if (input_tensor.first == wanted_shape) {
          (*node_shapes)[wanted_shape] = input_tensor.second.shape();
          is_input = true;
          break;
        } 
      }   
      input_count++;
      if (!is_input) {
        bool match= false;
        for (auto feeds_name : placeholder_name) {
          if (feeds_name.compare(wanted_shape) == 0) {
           match = true;
            break;
          }
          if (match)
            continue;
        } 
        output_tensor_names.push_back(wanted_shape);
      }   
    }     

  }
}
`
